core/src/main/scala-2/spire/syntax/Syntax.scala

```diff
@@ -261,13 +262,23 @@ trait ConvertableFromSyntax {
   implicit def convertableOps[A: ConvertableFrom](a: A): ConvertableFromOps[A] = new ConvertableFromOps(a)
 }
 
+@deprecated("Switch to fastFor", "0.18.0")
 trait CforSyntax {
```

Comment:
We should also add more information: it's not just a name change, `fastFor` _will_ have different semantics than `cfor`. What makes this particularly confusing, is it's hard to define what those semantics are exactly, and also if we don't implement exactly those same simplified semantics for Scala 2 it just adds to the confusion. This is extremely annoying and I'm honestly not sure what to do here.

@vasilmkd do you have any ideas about this? Basically, we want to replace `cfor` with `fastFor`, where `fastFor` has much simpler semantics (and thus will not be able to pass some of the current `cfor` tests). But so far, the only way Scala 2 users will find out about these simpler semantics is by compiling/running with Scala 3.

IMO the place to start is trying to define _exactly_ what `fastFor` does. I think it would be very helpful if we can define it in terms of ordinary Scala. In fact, we should make sure the "reference `fastFor`" and the macro `fastFor` pass the exact same test suite. Maybe something like this?
```scala
def referenceFastFor[A](init: A)(test: A => Boolean, next: A => A)(body: A => Unit): Unit =
  Iterator.iterate(init)(next).takeWhile(test).foreach(body)
```

Reply:
I agree with everything you've said in your comment, but I don't have anything smart to add. 😅

Reply:
I'll take your dumb ideas too 😝

Reply:
Btw, @cquiroz pointed out that at least one test for `cfor` compares to the non-macro implementation: https://github.com/cquiroz/spire/blob/dotty/tests/shared/src/test/scala-2/spire/syntax/CforSuite.scala

So maybe that is the difference between `cfor` and `fastFor`: `cfor` is like exactly if you write the `while` loop and inlined everything, whereas `fastFor` is just a fast version of that `Iterator.iterate ...` snippet I wrote above? If we can stick to this, I think it will make things so much easier to reason about.